# Description of the database

Let's start by reading the annotations.

```{r}
library(data.table)

seqs <- fread("db/data/medi_db/manifest.csv")
seqs[phylum == "", phylum := paste0("[", class, "]")]
seqs[, "taxa" := basename(filename)]
seqs[, sum(seqlength)/1e9]
seqs[, table(db, rank)]
```

And let's look at the matched foods:

```{r}
foods <- fread("db/data/dbs/food_matches.csv")
foods[, table(db)]
```

And have a look at the genome sizes.

```{r, fig.width=6, fig.height=3}
library(ggplot2)
theme_minimal() |> theme_set()

cn <- seqs[, .(N=.N, bps=sum(as.double(seqlength))), by=c("kingdom", "phylum", "db", "rank")]

ggplot(cn) +
  aes(x=phylum, y=bps, fill=rank) + 
  geom_bar(stat="identity", position="dodge", stroke=1, color="black") +
  facet_wrap(~ db) +
  scale_y_log10() +
  theme(axis.text.x = element_text(angle = 45, vjust=1, hjust=1)) +
  labs(x="", y="assembly size [bps]", fill="taxonomic rank")
ggsave("figures/db_size_summary.png", dpi=300)
```

```{r, fig.width=6, fig.height=3}
options(scipen=3)

ggplot(seqs) +
  aes(x=phylum, y=seqlength, color=rank) + 
  geom_jitter(width=0.4, stroke=0) +
  geom_boxplot(aes(group=phylum), color="black", width=0.2, outlier.color=NA) +
  facet_wrap(~ db) +
  scale_y_log10() +
  theme(axis.text.x = element_text(angle = 45, vjust=1, hjust=1)) +
  labs(x="", y="assembly size [bps]", color="taxonomic rank")
ggsave("figures/db_size_bps.pdf", width=6, height=3)
```

And the number of sequences.

```{r, fig.width=6, fig.height=3}
ggplot(cn) +
  aes(x=phylum, y=N, fill=rank) + 
  geom_bar(stat="identity", position="dodge", color="black") +
  facet_wrap(~ db) +
  theme(axis.text.x = element_text(angle = 45, vjust=1, hjust=1)) +
  scale_y_log10() +
  labs(x="", y="no. assemblies", fill="taxonomic rank")
ggsave("figures/db_size_n.pdf", width=6, height=3)
```

# Food taxonomy

Let's build up out food taxonomy and add in some metadata. We start by reading the food contents.

```{r}
contents <- fread("db/data/dbs/food_contents.csv.gz")[source_type == "Nutrient"]
contents[name=="Energy"]
```

Now we will start summarizing them over individual organisms.

```{r}
macros <- c("Proteins", "Carbohydrate", "Fiber (dietary)", "Fatty acids") 

broad_prepration <- function (classes) {
  tab <- table(classes)
  common <- names(tab)[which.min(tab)]
  allpreps <- paste0(unique(classes), collapse=",")
  if (grepl("raw", common)) {
    return("raw")
  }
  if (grepl("cooked", common)) {
    return("cooked")
  }
  if (grepl("oil", common)) {
    return("oil")
  }
  if (grepl("dried", common)) {
    return("dried or powder")
  }
  return("other")
}

collapsed <- contents[, .(
  amount=mean(standard_content, na.rm=T), 
  amount_sd=sd(standard_content, na.rm=T),
  preparation=broad_prepration(preparation_type)
  ), 
  by=c("compound_id", "matched_taxid", "name", "food_group", 
       "food_subgroup", "phylum", "genus", "species", "db", "rank")
]
collapsed[, preparation := factor(preparation, c("raw", "cooked", "dried or powder", "oil", "other"))]
collapsed <- collapsed[seqs, on="matched_taxid", nomatch=0]
collapsed[name %chin% macros, "relative" := amount / sum(amount), by="matched_taxid"]
uniq_foods <- unique(collapsed, by="matched_taxid")
uniq_foods <- cbind(uniq_foods$taxa, uniq_foods)
collapsed
```

Now we get the mash ANI estimates and convert them to a distance matrix.

```{r}
ani <- fread("db/data/mash_ani.csv")
taxa <- basename(names(ani))
ani <- as.matrix(ani)
rownames(ani) <- colnames(ani) <- taxa
D <- as.dist(1 - ani[seqs$taxa, seqs$taxa])
```

And we build the UPGMA tree.

```{r}
tree <- phangorn::upgma(D)
```

Now let's start to visualize.

```{r, fig.with=16, fig.height=16}
library(ggtree)
library(ggtreeExtra)

p <- ggtree(tree, layout="fan", open.angle=20) %<+% cbind(seqs$taxa, seqs) |> rotate_tree(90)
p <- p + geom_tippoint(aes(color=phylum))
```

Let's start to annotate with some more things.

```{r, fig.width=18, fig.height=18}
p + 
  geom_fruit(data=uniq_foods, aes(y=taxa, x=0, shape=preparation), geom="geom_point", size=1.1) +
  geom_fruit(
    data=collapsed[name %chin% macros], 
    aes(y=taxa, x=name, fill=name, alpha=relative),
    color="gray50",
    geom="geom_tile",
    pwidth=0.07
  ) +
  geom_fruit(
    data=collapsed[name == "Energy"],
    aes(y=taxa, x=amount),
    fill="black",
    geom=geom_bar,
    stat="identity",
    position = position_identityx(),
    pwidth=0.2,
    axis.params=list(
      axis       = "x",
      text.size  = 3,
      hjust      = 0,
      vjust      = 0.5,
      nbreak     = 4,
    ),
    grid.params=list()
  )
ggsave("figures/food_tree.pdf", width=18, height=18)
```

Association with nutrients:

```{r}
library(vegan)

for (m in macros) {
  print(m)
  df <- collapsed[name == m] |> as("data.frame")
  a <- ani[df$taxa, df$taxa]
  p <- adonis2(dist(1 - a) ~ amount, data=df)
  print(p)
}
```

Macronutrient composition distance vs ANI

```{r}
macro_abs <- collapsed[!grepl("Energy", name) & db=="genbank"] |> dcast(taxa ~ name, fill=0, value.var = "amount", fun.aggregate = mean)
taxa <- macro_abs[, taxa]
macro_abs <- as.matrix(macro_abs[, "taxa" := NULL])
rownames(macro_abs) <- taxa
D_macros <- dist(macro_abs)
```

```{r}
dt <- data.table(divergence=as.numeric(as.dist(1 - ani[taxa, taxa])), macro_dist=as.numeric(D_macros))[divergence < 1]
mantel(D_macros, as.dist(1-ani[taxa, taxa]), method="spear")
ggplot(dt) +
  aes(x=divergence, y=macro_dist/1000) +
  geom_point(stroke=0, size=1, alpha=0.5) +
  stat_smooth(aes(x=divergence), method="gam") +
  labs(x="1-ANI", y="macronutrient distance [g/100g]")
ggsave("figures/ani_vs_macros.png", width=6, height=4, dpi=300)
```

## Mapping speed

```{r}
library(stringr)
pattern <- "(\\d+) sequences \\(([\\d\\.]+) Mbp\\) processed in ([\\d\\.]+)s"

speeds <- function(path) {
  lines <- readLines(con = path)
  matches <- str_match(lines, pattern)
  matches <- apply(matches[!is.na(matches[, 1]), 2:4], 2, as.numeric)
  matches <- as.data.table(matches)
  names(matches) <- c("nreads", "size_mbps", "time_seconds")
  matches[, "path" := path]
  matches[, "iteration" := 1:.N]
  return(matches)
}
```

```{r, fig.height=4, fig.width=5}
library(patchwork)

measures <- rbind(
  speeds("metacardis/data/kraken_batch1.log"),
  speeds("metacardis/data/kraken_batch2.log")
)

timeplot <- ggplot(measures) +
  aes(x=iteration, y=time_seconds/nreads*1e6, color=path, group=path) +
  geom_point() +
  scale_y_log10() +
  labs(y="time per 1M reads [s]") +
  guides(color="none")

rateplot <- ggplot(measures[iteration > 0]) +
  aes(x=nreads/time_seconds, fill=path, group=path) +
  stat_density(position="stack") +
    geom_vline(aes(xintercept=median(nreads/time_seconds))) +
  guides(fill="none") + labs(x="classification rate [reads/second]")

timeplot / rateplot
ggsave("figures/kraken_benchmarks.pdf", width=5, height=4)
```

## Write figure data

```{r}
fig_data <- list(
  `Fig. 1A/B` = seqs,
  `Fig. 1C` = collapsed
)

si_data <- list(
  `Fig. S1A` = dt,
  `Fig. S1B` = measures,
  `Fig. S1C` = fread("figure_data/ihmp_genome_size.csv")
)
```

```{r}
library(openxlsx)
write.xlsx(fig_data, "figure_data/fig1.xlsx")
write.xlsx(si_data, "figure_data/figS1.xlsx")
```

DB Stats

```{r}
contents <- fread("db/data/dbs/food_contents.csv.gz")
```

```{r}
summ <- contents[, .(
  median_content = median(standard_content), 
  min_content = min(standard_content),
  max_content = max(standard_content),
  iqr_content=iqr(standard_content), 
  n_foods_measured=uniqueN(food_id)
  ),
  by = c("compound_id", "name", "mono_mass", "source_type", "unit")               
][, "n_total_foods" := nrow(foods)]
fwrite(summ, "db/data/content_summaries.csv")
summ
```